common/src/main/java/com/tc/net/core/TCConnectionImpl.java

```diff
@@ -402,50 +410,75 @@ private int doWriteToBuffer() throws IOException {
     }
   }
 
-  private void buildWriteContextsFromMessages() {
-    TCNetworkMessage messagesToWrite[];
-    synchronized (this.writeMessages) {
-      if (this.closed.isSet()) { return; }
-      messagesToWrite = this.writeMessages.toArray(new TCNetworkMessage[this.writeMessages.size()]);
-      this.writeMessages.clear();
+  private boolean buildWriteContextsFromMessages(boolean failfast) {
+    try {
+      if (failfast) {
+        if (!writeContextControl.tryAcquire()) {
+          return false;
+        }
+      } else {
+        writeContextControl.acquire();
+      }
+    } catch (InterruptedException ie) {
+      Thread.currentThread().interrupt();
+      return false;
     }
-    ArrayList<TCNetworkMessage> currentBatch = (MSG_GROUPING_ENABLED
-        ? new ArrayList<>()
-        : null);
     
+    this.writeContexts.removeIf(WriteContext::isNotValid);
```

Comment:
this is here for GC.  I don't think we should call it a ton because it is probably expensive if the list gets long.  Anything that goes invalid before being removed will just get dropped on the floor because prepareToSend returns false when there the entire group is unable to commit.  maybe we only call this when fail fast is true and unable to acquire the lock?